Botania, JEI and Crafting Station crash

My crafting station was placed beside the petal apothecary from Botania and it recognized the internal slots as a chest in the GUI.

Since I couldn't click to insert or take items from the slots I just assumed it was fine since it respected the isValidItem/canTake methods on those slots.

But then I tried to click the plus sign on a recipe in JEI and since the crafring station wasnt empty it tried to move the items in to the slots of the petal apothecary and since those items should never be in there it crashed the game.

Hopefully a not too confusing screenshot:
![2016-11-16_01 21 05](https://cloud.githubusercontent.com/assets/2662937/20330050/63814c76-ab9c-11e6-8a4b-3036f69ca4cc.png)

Now afaik JEI is at fault but I reported it in all three repos:
https://github.com/SlimeKnights/TinkersConstruct/issues/2570
https://github.com/mezz/JustEnoughItems/issues/615
https://github.com/Vazkii/Botania/issues/2036

```
net.minecraft.util.ReportedException: Ticking block entity
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:715) ~[MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:613) ~[MinecraftServer.class:?]
	at net.minecraft.server.integrated.IntegratedServer.func_71217_p(IntegratedServer.java:149) ~[bzl.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:471) [MinecraftServer.class:?]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_25]
Caused by: java.lang.ClassCastException: net.minecraft.item.Item cannot be cast to vazkii.botania.api.recipe.IFlowerComponent
	at vazkii.botania.common.block.tile.TileAltar.func_73660_a(TileAltar.java:263) ~[TileAltar.class:?]
	at net.minecraft.world.World.func_72939_s(World.java:1804) ~[aid.class:?]
	at net.minecraft.world.WorldServer.func_72939_s(WorldServer.java:619) ~[ls.class:?]
	at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:709) ~[MinecraftServer.class:?]
	... 4 more
```

Relevant versions:
- Forge{12.18.2.2125} [Minecraft Forge] (forge-1.10.2-12.18.2.2125.jar) 
- mantle{1.10.2-1.1.1.194} [Mantle] (Mantle-1.10.2-1.1.1.jar) 
- tconstruct{1.10.2-2.5.6b.jenkins424} [Tinkers' Construct] (TConstruct-1.10.2-2.5.6b.jar) 
- JEI{3.13.3.372} [Just Enough Items] (jei_1.10.2-3.13.3.372.jar) 
- Botania{r1.9-337} [Botania] (Botania r1.9-337.jar)

--- a/src/main/java/slimeknights/tconstruct/TConstruct.java
+++ b/src/main/java/slimeknights/tconstruct/TConstruct.java
@@ -60,7 +60,8 @@ import slimeknights.tconstruct.world.TinkerWorld;
     version = TConstruct.modVersion,
     guiFactory = "slimeknights.tconstruct.common.config.ConfigGui$ConfigGuiFactory",
     dependencies = "required-after:Forge@[12.18.1.2073,);"
-                   + "required-after:mantle@[1.10.2-1.0.0,)",
+                   + "required-after:mantle@[1.10.2-1.0.0,);"
+                   + "after:JEI@[3.13.6.387,)",
     acceptedMinecraftVersions = "[1.10.2, 1.11)")
 public class TConstruct {
 

--- a/src/main/java/slimeknights/tconstruct/gadgets/item/ItemBlockRack.java
+++ b/src/main/java/slimeknights/tconstruct/gadgets/item/ItemBlockRack.java
@@ -4,14 +4,12 @@ import net.minecraft.block.Block;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemMultiTexture;
 import net.minecraft.item.ItemStack;
-import net.minecraft.nbt.NBTTagCompound;
-
 import java.util.List;
 
 import javax.annotation.Nonnull;
 
 import slimeknights.tconstruct.library.Util;
-import slimeknights.tconstruct.shared.tileentity.TileTable;
+import slimeknights.tconstruct.tools.common.item.ItemBlockTable;
 
 public class ItemBlockRack extends ItemMultiTexture {
 
@@ -22,9 +20,8 @@ public class ItemBlockRack extends ItemMultiTexture {
   @Override
   public void addInformation(@Nonnull ItemStack stack, @Nonnull EntityPlayer playerIn, @Nonnull List<String> tooltip, boolean advanced) {
     if(stack.hasTagCompound()) {
-      NBTTagCompound tag = stack.getTagCompound().getCompoundTag(TileTable.FEET_TAG);
-      ItemStack legs = ItemStack.loadItemStackFromNBT(tag);
-      if(legs != null && legs.getItem() != null) {
+      ItemStack legs = ItemBlockTable.getLegStack(stack);
+      if(legs != null) {
         tooltip.add(legs.getDisplayName());
       }
 

--- a/src/main/java/slimeknights/tconstruct/library/client/GuiUtil.java
+++ b/src/main/java/slimeknights/tconstruct/library/client/GuiUtil.java
@@ -153,6 +153,22 @@ public class GuiUtil {
     }
   }
 
+  public static FluidStack getFluidStackAtPosition(SmelteryTank tank, int mouseX, int mouseY, int xmin, int ymin, int xmax, int ymax) {
+    if(xmin <= mouseX && mouseX < xmax && ymin <= mouseY && mouseY < ymax) {
+      int[] heights = calcLiquidHeights(tank.getFluids(), tank.getCapacity(), ymax - ymin);
+      int y = ymax - mouseY - 1;
+
+      for(int i = 0; i < heights.length; i++) {
+        if(y < heights[i]) {
+          return tank.getFluids().get(i);
+        }
+        y -= heights[i];
+      }
+    }
+
+    return null;
+  }
+
   public static void handleTankClick(SmelteryTank tank, int mouseX, int mouseY, int xmin, int ymin, int xmax, int ymax) {
     if(xmin <= mouseX && mouseX < xmax && ymin <= mouseY && mouseY < ymax) {
       int[] heights = calcLiquidHeights(tank.getFluids(), tank.getCapacity(), ymax - ymin);

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/CastingRecipeWrapper.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/CastingRecipeWrapper.java
@@ -7,7 +7,6 @@ import net.minecraft.item.ItemStack;
 import net.minecraftforge.fluids.FluidStack;
 
 import java.awt.*;
-import java.util.Collections;
 import java.util.List;
 
 import javax.annotation.Nonnull;
@@ -64,7 +63,7 @@ public class CastingRecipeWrapper extends BlankRecipeWrapper {
 
   @Override
   public void getIngredients(IIngredients ingredients) {
-    ingredients.setInputs(ItemStack.class, cast);
+    ingredients.setInputLists(ItemStack.class, ImmutableList.of(cast));
     ingredients.setInputs(FluidStack.class, inputFluid);
     ingredients.setOutputs(ItemStack.class, lazyInitOutput());
   }

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/CraftingStationRecipeTransferInfo.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/CraftingStationRecipeTransferInfo.java
@@ -42,7 +42,10 @@ public class CraftingStationRecipeTransferInfo implements IRecipeTransferInfo<Co
   @Override
   public List<Slot> getInventorySlots(ContainerCraftingStation container) {
     List<Slot> slots = new ArrayList<Slot>();
-    for(int i = 10; i < container.inventorySlots.size(); i++) {
+
+    // we skip all slots from within the side inventory and crafting grid for transfer
+    // side inventory can cause too many issues since transfer is not validated the same way as clicking does
+    for(int i = container.getPlayerInventoryStart(); i < container.inventorySlots.size(); i++) {
       slots.add(container.getSlot(i));
     }
     return slots;

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/Focus.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/Focus.java
@@ -1,0 +1,24 @@
+package slimeknights.tconstruct.plugin.jei;
+
+import mezz.jei.api.recipe.IFocus;
+
+// this is simply a private copy since the JEI implementation is not in the API
+public class Focus<V> implements IFocus<V> {
+  private final Mode mode;
+  private final V value;
+
+  public Focus(Mode mode, V value) {
+    this.mode = mode;
+    this.value = value;
+  }
+
+  @Override
+  public V getValue() {
+    return value;
+  }
+
+  @Override
+  public Mode getMode() {
+    return mode;
+  }
+}

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/JEIPlugin.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/JEIPlugin.java
@@ -4,6 +4,7 @@ import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 
+import net.minecraft.client.gui.inventory.GuiContainer;
 import net.minecraft.init.Blocks;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
@@ -15,6 +16,7 @@ import java.util.List;
 import java.util.Map;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 
 import mezz.jei.api.IGuiHelper;
 import mezz.jei.api.IJeiHelpers;
@@ -22,6 +24,7 @@ import mezz.jei.api.IJeiRuntime;
 import mezz.jei.api.IModPlugin;
 import mezz.jei.api.IModRegistry;
 import mezz.jei.api.ISubtypeRegistry;
+import mezz.jei.api.gui.BlankAdvancedGuiHandler;
 import mezz.jei.api.ingredients.IModIngredientRegistration;
 import mezz.jei.api.recipe.VanillaRecipeCategoryUid;
 import slimeknights.tconstruct.TConstruct;
@@ -33,6 +36,9 @@ import slimeknights.tconstruct.library.smeltery.ICastingRecipe;
 import slimeknights.tconstruct.shared.block.BlockTable;
 import slimeknights.tconstruct.smeltery.TinkerSmeltery;
 import slimeknights.tconstruct.smeltery.block.BlockCasting;
+import slimeknights.tconstruct.smeltery.client.GuiSmeltery;
+import slimeknights.tconstruct.smeltery.client.GuiTinkerTank;
+import slimeknights.tconstruct.smeltery.client.IGuiLiquidTank;
 import slimeknights.tconstruct.tools.TinkerTools;
 import slimeknights.tconstruct.tools.common.block.BlockToolTable;
 
@@ -122,6 +128,11 @@ public class JEIPlugin implements IModPlugin {
           registry.addRecipes(ImmutableList.of(new CastingRecipeWrapper(recipe, castingCategory.castingBasin)));
         }
       }
+
+      // liquid recipe lookup for smeltery and tinker tank
+      registry.addAdvancedGuiHandlers(
+          new TinkerGuiTankHandler<GuiTinkerTank>(GuiTinkerTank.class),
+          new TinkerGuiTankHandler<GuiSmeltery>(GuiSmeltery.class));
     }
 
     // drying rack
@@ -136,5 +147,26 @@ public class JEIPlugin implements IModPlugin {
 
   @Override
   public void onRuntimeAvailable(@Nonnull IJeiRuntime jeiRuntime) {
+  }
+
+
+  private static class TinkerGuiTankHandler<T extends GuiContainer & IGuiLiquidTank> extends BlankAdvancedGuiHandler<T> {
+    private Class<T> clazz;
+
+    public TinkerGuiTankHandler(Class<T> clazz) {
+      this.clazz = clazz;
+    }
+
+    @Nonnull
+    @Override
+    public Class<T> getGuiContainerClass() {
+      return clazz;
+    }
+
+    @Nullable
+    @Override
+    public Object getIngredientUnderMouse(T guiContainer, int mouseX, int mouseY) {
+      return guiContainer.getFluidStackAtPosition(mouseX, mouseY);
+    }
   }
 }

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/JEIPlugin.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/JEIPlugin.java
@@ -25,6 +25,7 @@ import mezz.jei.api.IModPlugin;
 import mezz.jei.api.IModRegistry;
 import mezz.jei.api.ISubtypeRegistry;
 import mezz.jei.api.gui.BlankAdvancedGuiHandler;
+import mezz.jei.api.gui.ICraftingGridHelper;
 import mezz.jei.api.ingredients.IModIngredientRegistration;
 import mezz.jei.api.recipe.VanillaRecipeCategoryUid;
 import slimeknights.tconstruct.TConstruct;
@@ -33,6 +34,7 @@ import slimeknights.tconstruct.library.TinkerRegistry;
 import slimeknights.tconstruct.library.smeltery.Cast;
 import slimeknights.tconstruct.library.smeltery.CastingRecipe;
 import slimeknights.tconstruct.library.smeltery.ICastingRecipe;
+import slimeknights.tconstruct.library.tools.IToolPart;
 import slimeknights.tconstruct.shared.block.BlockTable;
 import slimeknights.tconstruct.smeltery.TinkerSmeltery;
 import slimeknights.tconstruct.smeltery.block.BlockCasting;
@@ -46,10 +48,44 @@ import slimeknights.tconstruct.tools.common.block.BlockToolTable;
 public class JEIPlugin implements IModPlugin {
 
   public static IJeiHelpers jeiHelpers;
+  // crafting grid slots, integer constants from the default crafting grid implementation
+  private static final int craftOutputSlot = 0;
+  private static final int craftInputSlot1 = 1;
+  public static ICraftingGridHelper craftingGridHelper;
 
   @Override
-  public void registerItemSubtypes(ISubtypeRegistry subtypeRegistry) {
+  public void registerItemSubtypes(ISubtypeRegistry registry) {
+    TableSubtypeInterpreter tableInterpreter = new TableSubtypeInterpreter();
+    PatternSubtypeInterpreter patternInterpreter = new PatternSubtypeInterpreter();
 
+    // drying racks and item racks
+    if(TConstruct.pulseManager.isPulseLoaded(TinkerGadgets.PulseId)) {
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerGadgets.rack), tableInterpreter);
+    }
+
+    // tools
+    if(TConstruct.pulseManager.isPulseLoaded(TinkerTools.PulseId)) {
+      // tool tables
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerTools.toolTables), tableInterpreter);
+      registry.registerSubtypeInterpreter(Item.getItemFromBlock(TinkerTools.toolForge), tableInterpreter);
+
+      // tool parts
+      ToolPartSubtypeInterpreter toolPartInterpreter = new ToolPartSubtypeInterpreter();
+      for(IToolPart part : TinkerRegistry.getToolParts()) {
+        if(part instanceof Item) {
+          registry.registerSubtypeInterpreter((Item)part, toolPartInterpreter);
+        }
+      }
+
+      // tool patterns
+      registry.registerSubtypeInterpreter(TinkerTools.pattern, patternInterpreter);
+    }
+
+    // casts
+    if(TConstruct.pulseManager.isPulseLoaded(TinkerSmeltery.PulseId)) {
+      registry.registerSubtypeInterpreter(TinkerSmeltery.cast, patternInterpreter);
+      registry.registerSubtypeInterpreter(TinkerSmeltery.clayCast, patternInterpreter);
+    }
   }
 
   @Override
@@ -61,6 +97,9 @@ public class JEIPlugin implements IModPlugin {
   public void register(@Nonnull IModRegistry registry) {
     jeiHelpers = registry.getJeiHelpers();
     IGuiHelper guiHelper = jeiHelpers.getGuiHelper();
+
+    // crafting helper used by the shaped table wrapper
+    craftingGridHelper = guiHelper.createCraftingGridHelper(craftInputSlot1, craftOutputSlot);
 
     if(TConstruct.pulseManager.isPulseLoaded(TinkerTools.PulseId)) {
       // crafting table shiftclicking

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/PatternSubtypeInterpreter.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/PatternSubtypeInterpreter.java
@@ -1,0 +1,23 @@
+package slimeknights.tconstruct.plugin.jei;
+
+import mezz.jei.api.ISubtypeRegistry.ISubtypeInterpreter;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import slimeknights.tconstruct.library.tools.Pattern;
+
+// Handles pattern and cast subtypes
+public class PatternSubtypeInterpreter implements ISubtypeInterpreter {
+
+  @Override
+  public String getSubtypeInfo(ItemStack stack) {
+    String meta = stack.getItemDamage() + ":";
+
+    Item part = Pattern.getPartFromTag(stack);
+    if(part == null) {
+      return meta;
+    }
+
+    return meta + part.getRegistryName();
+  }
+
+}

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/TableRecipeWrapper.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/TableRecipeWrapper.java
@@ -13,14 +13,20 @@ import java.util.List;
 
 import javax.annotation.Nonnull;
 
+import mezz.jei.api.gui.IGuiIngredientGroup;
+import mezz.jei.api.gui.IGuiItemStackGroup;
+import mezz.jei.api.gui.IRecipeLayout;
 import mezz.jei.api.ingredients.IIngredients;
 import mezz.jei.api.recipe.BlankRecipeWrapper;
+import mezz.jei.api.recipe.IFocus;
 import mezz.jei.api.recipe.IStackHelper;
+import mezz.jei.api.recipe.wrapper.ICustomCraftingRecipeWrapper;
 import mezz.jei.api.recipe.wrapper.IShapedCraftingRecipeWrapper;
 import slimeknights.tconstruct.shared.block.BlockTable;
 import slimeknights.tconstruct.tools.common.TableRecipe;
+import slimeknights.tconstruct.tools.common.item.ItemBlockTable;
 
-public class TableRecipeWrapper extends BlankRecipeWrapper implements IShapedCraftingRecipeWrapper {
+public class TableRecipeWrapper extends BlankRecipeWrapper implements IShapedCraftingRecipeWrapper, ICustomCraftingRecipeWrapper {
 
   private final TableRecipe recipe;
   private final int width;
@@ -73,7 +79,9 @@ public class TableRecipeWrapper extends BlankRecipeWrapper implements IShapedCra
     }
   }
 
+  @SuppressWarnings("rawtypes")
   @Override
+  @Deprecated
   public List getInputs() {
     return Arrays.asList(recipe.getInput());
   }
@@ -92,5 +100,67 @@ public class TableRecipeWrapper extends BlankRecipeWrapper implements IShapedCra
   @Override
   public int getHeight() {
     return height;
+  }
+
+  private boolean isOutputBlock(ItemStack stack) {
+    if(stack == null) {
+      return false;
+    }
+
+    for(ItemStack output : recipe.outputBlocks) {
+      // if the item matches the oredict entry, it is an output block
+      if(OreDictionary.itemMatches(output, stack, false)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayout recipeLayout, IIngredients ingredients) {
+    IGuiItemStackGroup guiItemStacks = recipeLayout.getItemStacks();
+
+    List<List<ItemStack>> inputs = ingredients.getInputs(ItemStack.class);
+    List<ItemStack> outputs = ingredients.getOutputs(ItemStack.class);
+
+    // determine the focused stack
+    IFocus<?> ifocus = recipeLayout.getFocus();
+    Object focusObj = ifocus.getValue();
+
+    // if the thing in focus is an itemstack
+    if(focusObj instanceof ItemStack) {
+      IGuiIngredientGroup<ItemStack> guiIngredients = recipeLayout.getIngredientsGroup(ItemStack.class);
+      ItemStack focus = (ItemStack)focusObj;
+      IFocus.Mode mode = ifocus.getMode();
+
+      // input means we clicked on an ingredient, make sure it is one that affects the legs
+      if(mode == IFocus.Mode.INPUT && isOutputBlock(focus)) {
+        // first, get the output recipe
+        ItemStack output = recipe.getPlainRecipeOutput();
+        BlockTable block = (BlockTable) Block.getBlockFromItem(output.getItem());
+
+        // then create a stack with the focus item (which we already validated above)
+        ItemStack outputFocus = BlockTable.createItemstack(block, output.getItemDamage(), Block.getBlockFromItem(focus.getItem()),
+            focus.getItemDamage());
+
+        // and finally, set the focus override for the recipe
+        guiIngredients.setOverrideDisplayFocus(new Focus<ItemStack>(IFocus.Mode.OUTPUT, outputFocus));
+      }
+
+      // if we clicked the table, remove all items which affect the legs textures that are not the leg item
+      else if(mode == IFocus.Mode.OUTPUT) {
+        // so determine the legs
+        ItemStack legs = ItemBlockTable.getLegStack(focus);
+        if(legs != null) {
+          // and loop through all slots removing leg affecting inputs which don't match
+          guiIngredients.setOverrideDisplayFocus(new Focus<ItemStack>(IFocus.Mode.INPUT, legs));
+        }
+      }
+    }
+
+    // add the itemstacks to the grid
+    JEIPlugin.craftingGridHelper.setInputStacks(guiItemStacks, inputs, this.getWidth(), this.getHeight());
+    JEIPlugin.craftingGridHelper.setOutput(guiItemStacks, outputs);
   }
 }

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/TableSubtypeInterpreter.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/TableSubtypeInterpreter.java
@@ -1,0 +1,28 @@
+package slimeknights.tconstruct.plugin.jei;
+
+import mezz.jei.api.ISubtypeRegistry.ISubtypeInterpreter;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import slimeknights.tconstruct.library.utils.TagUtil;
+import slimeknights.tconstruct.shared.tileentity.TileTable;
+
+// Hanldes table and rack subtypes
+public class TableSubtypeInterpreter implements ISubtypeInterpreter {
+
+  @Override
+  public String getSubtypeInfo(ItemStack stack) {
+    // we have to handle the metadata here
+    String meta = stack.getMetadata() + ":";
+
+    // if the legs exist, return that for the identification key
+    NBTTagCompound tag = TagUtil.getTagSafe(stack).getCompoundTag(TileTable.FEET_TAG);
+    ItemStack legs = ItemStack.loadItemStackFromNBT(tag);
+    if(legs != null && legs.getItem() != null) {
+      return meta + legs.getItem().getRegistryName() + ":" + legs.getMetadata();
+    }
+
+    // otherwise, simply go back to the meta
+    return meta;
+  }
+
+}

--- a/src/main/java/slimeknights/tconstruct/plugin/jei/ToolPartSubtypeInterpreter.java
+++ b/src/main/java/slimeknights/tconstruct/plugin/jei/ToolPartSubtypeInterpreter.java
@@ -1,0 +1,23 @@
+package slimeknights.tconstruct.plugin.jei;
+
+import mezz.jei.api.ISubtypeRegistry.ISubtypeInterpreter;
+import net.minecraft.item.ItemStack;
+import slimeknights.tconstruct.library.materials.Material;
+import slimeknights.tconstruct.library.utils.TinkerUtil;
+
+// Handles all Tinker tool parts subtypes
+public class ToolPartSubtypeInterpreter implements ISubtypeInterpreter {
+
+  @Override
+  public String getSubtypeInfo(ItemStack stack) {
+    String meta = stack.getItemDamage() + ":";
+
+    Material material = TinkerUtil.getMaterialFromStack(stack);
+    if(material == null) {
+      return meta;
+    }
+
+    return meta + material.getIdentifier();
+  }
+
+}

--- a/src/main/java/slimeknights/tconstruct/smeltery/client/GuiSmeltery.java
+++ b/src/main/java/slimeknights/tconstruct/smeltery/client/GuiSmeltery.java
@@ -2,6 +2,8 @@ package slimeknights.tconstruct.smeltery.client;
 
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
+
 import java.io.IOException;
 import java.util.List;
 
@@ -13,7 +15,7 @@ import slimeknights.tconstruct.smeltery.inventory.ContainerSmeltery;
 import slimeknights.tconstruct.smeltery.tileentity.TileSmeltery;
 import slimeknights.tconstruct.tools.common.inventory.ContainerSideInventory;
 
-public class GuiSmeltery extends GuiHeatingStructureFuelTank {
+public class GuiSmeltery extends GuiHeatingStructureFuelTank implements IGuiLiquidTank {
 
   public static final ResourceLocation BACKGROUND = Util.getResource("textures/gui/smeltery.png");
 
@@ -91,6 +93,11 @@ public class GuiSmeltery extends GuiHeatingStructureFuelTank {
       GuiUtil.handleTankClick(smeltery.getTank(), mouseX - cornerX, mouseY - cornerY, 8, 16, 60, 68);
     }
     super.mouseClicked(mouseX, mouseY, mouseButton);
+  }
+
+  @Override
+  public FluidStack getFluidStackAtPosition(int mouseX, int mouseY) {
+    return GuiUtil.getFluidStackAtPosition(smeltery.getTank(), mouseX - cornerX, mouseY - cornerY, 8, 16, 60, 68);
   }
 
 }

--- a/src/main/java/slimeknights/tconstruct/smeltery/client/GuiTinkerTank.java
+++ b/src/main/java/slimeknights/tconstruct/smeltery/client/GuiTinkerTank.java
@@ -7,13 +7,14 @@ import net.minecraft.client.gui.inventory.GuiContainer;
 import net.minecraft.client.renderer.GlStateManager;
 import net.minecraft.inventory.Container;
 import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.fluids.FluidStack;
 import slimeknights.mantle.client.gui.GuiElement;
 import slimeknights.mantle.inventory.BaseContainer;
 import slimeknights.tconstruct.library.Util;
 import slimeknights.tconstruct.library.client.GuiUtil;
 import slimeknights.tconstruct.smeltery.tileentity.TileTinkerTank;
 
-public class GuiTinkerTank extends GuiContainer {
+public class GuiTinkerTank extends GuiContainer implements IGuiLiquidTank {
 
   public static final ResourceLocation BACKGROUND = Util.getResource("textures/gui/tinker_tank.png");
   protected GuiElement scala = new GuiElement(122, 0, 106, 106, 256, 256);
@@ -64,6 +65,11 @@ public class GuiTinkerTank extends GuiContainer {
     }
 
     super.mouseClicked(mouseX, mouseY, mouseButton);
+  }
+
+  @Override
+  public FluidStack getFluidStackAtPosition(int mouseX, int mouseY) {
+    return GuiUtil.getFluidStackAtPosition(tinkerTank.getTank(), mouseX - guiLeft, mouseY - guiTop, 8, 16, 114, 122);
   }
 
   protected void drawBackground(ResourceLocation background) {

--- a/src/main/java/slimeknights/tconstruct/smeltery/client/IGuiLiquidTank.java
+++ b/src/main/java/slimeknights/tconstruct/smeltery/client/IGuiLiquidTank.java
@@ -1,0 +1,7 @@
+package slimeknights.tconstruct.smeltery.client;
+
+import net.minecraftforge.fluids.FluidStack;
+
+public interface IGuiLiquidTank {
+  FluidStack getFluidStackAtPosition(int mouseX, int mouseY);
+}

--- a/src/main/java/slimeknights/tconstruct/tools/common/TableRecipe.java
+++ b/src/main/java/slimeknights/tconstruct/tools/common/TableRecipe.java
@@ -55,4 +55,11 @@ public class TableRecipe extends ShapedOreRecipe {
     }
     return super.getRecipeOutput();
   }
+
+  /**
+   * Gets the recipe output without applying the legs block
+   */
+  public ItemStack getPlainRecipeOutput() {
+    return output;
+  }
 }

--- a/src/main/java/slimeknights/tconstruct/tools/common/inventory/ContainerCraftingStation.java
+++ b/src/main/java/slimeknights/tconstruct/tools/common/inventory/ContainerCraftingStation.java
@@ -120,4 +120,10 @@ public class ContainerCraftingStation extends ContainerTinkerStation<TileCraftin
     return null;
   }
 
+  /**
+   * @return the starting slot for the player inventory. Present for usage in the JEI crafting station support
+   */
+  public int getPlayerInventoryStart() {
+    return playerInventoryStart;
+  }
 }

--- a/src/main/java/slimeknights/tconstruct/tools/common/item/ItemBlockTable.java
+++ b/src/main/java/slimeknights/tconstruct/tools/common/item/ItemBlockTable.java
@@ -16,6 +16,7 @@ import slimeknights.mantle.util.LocUtils;
 import slimeknights.tconstruct.library.Util;
 import slimeknights.tconstruct.library.smeltery.ICast;
 import slimeknights.tconstruct.library.tools.IPattern;
+import slimeknights.tconstruct.library.utils.TagUtil;
 import slimeknights.tconstruct.shared.tileentity.TileTable;
 import slimeknights.tconstruct.tools.common.block.BlockToolTable;
 
@@ -32,15 +33,31 @@ public class ItemBlockTable extends ItemBlockMeta {
       return;
     }
 
-    NBTTagCompound tag = stack.getTagCompound().getCompoundTag(TileTable.FEET_TAG);
-    ItemStack legs = ItemStack.loadItemStackFromNBT(tag);
-    if(legs != null && legs.getItem() != null) {
+    ItemStack legs = getLegStack(stack);
+    if(legs != null) {
       tooltip.add(legs.getDisplayName());
     }
 
     if(stack.getTagCompound().hasKey("inventory")) {
       this.addInventoryInformation(stack, playerIn, tooltip, advanced);
     }
+  }
+
+  /**
+   * Gets the itemstack that determines the leg's texture from the table
+   * @param table  Input table
+   * @return  The itemstack determining the leg's texture, or null if none exists
+   */
+  public static ItemStack getLegStack(ItemStack table) {
+    NBTTagCompound tag = TagUtil.getTagSafe(table).getCompoundTag(TileTable.FEET_TAG);
+    ItemStack stack = ItemStack.loadItemStackFromNBT(tag);
+
+    // don't use a stack with a null item
+    if(stack == null || stack.getItem() == null) {
+      return null;
+    }
+
+    return stack;
   }
 
   protected void addInventoryInformation(@Nonnull ItemStack stack, @Nonnull EntityPlayer playerIn, @Nonnull List<String> tooltip, boolean advanced) {

